app/org/sagebionetworks/bridge/dynamodb/DynamoIndexHelper.java

```diff
@@ -0,0 +1,97 @@
+ * class also encapsulates logic to re-query tables to get full table entries.
+ */
+public class DynamoIndexHelper {
+    private Index index;
+    private DynamoDBMapper mapper;
+
+    /** DynamoDB index. THis is used to query the global secondary index. This is configured by Spring. */
+    public void setIndex(Index index) {
+        this.index = index;
+    }
+
+    /**
+     * DynamoDB mapper. This is used to re-query the DynamoDB table to get full entries from the key objects. This is
+     * configured by Spring.
+     */
+    public void setMapper(DynamoDBMapper mapper) {
+        this.mapper = mapper;
+    }
+
+    /**
+     * Queries the global secondary index with the specified key name and value. Results will be returned as a list of
+     * the specified class.
+     *
+     * @param clazz
+     *         expected result class
+     * @param indexKeyName
+     *         index key name to query on
+     * @param indexKeyValue
+     *         index key value to query on
+     * @param <T>
+     *         expected result type
+     * @return list of query results
+     */
+    @SuppressWarnings("unchecked")
+    public <T> List<T> query(@Nonnull Class<? extends T> clazz, @Nonnull String indexKeyName,
+            @Nonnull Object indexKeyValue) {
+        // query the index
+        Iterable<Item> itemIter = queryHelper(indexKeyName, indexKeyValue);
+
+        // In general, we only project keys onto global secondary indices, to save storage space. This means the
+        // objects we get back aren't full fledged objects. However, we can use them as "key objects" to re-query
+        // the DDB table to get full results.
+        //
+        // Also, for some reason, batchLoad() takes a List<Object>, not a List<T> or List<?>.
+        List<Object> recordKeyList = new ArrayList<>();
+        for (Item oneItem : itemIter) {
+            T oneRecord = BridgeObjectMapper.get().convertValue(oneItem.asMap(), clazz);
+            recordKeyList.add(oneRecord);
+        }
+
+        // Using the "key objects", batch query DDB to get full records. For some reason, batchLoad() returns a map.
+        // Flatten that map into a list.
+        Map<String, List<Object>> resultMap = mapper.batchLoad(recordKeyList);
+        List<T> recordList = new ArrayList<>();
+        for (List<Object> resultList : resultMap.values()) {
+            for (Object oneResult : resultList) {
+                if (!clazz.isInstance(oneResult)) {
+                    // This should never happen, but just in case.
+                    throw new BridgeServiceException(String.format(
+                            "DynamoDB returned objects of type %s instead of %s",
+                            oneResult.getClass().getName(), clazz.getName()));
+                }
+
+                recordList.add((T) oneResult);
+            }
+        }
+
+        return recordList;
+    }
+
+    /**
```

Comment:
Discussed this offline with Alx. I felt this was important enough to include in the javadoc.